onnxruntime/core/providers/mkldnn/mkldnn_execution_provider.h

```diff
@@ -37,6 +44,23 @@ class MKLDNNExecutionProvider : public IExecutionProvider {
   }
 
   virtual std::shared_ptr<KernelRegistry> GetKernelRegistry() const override;
+
+  std::shared_ptr<mkldnn::memory> GetWeightsMemory(std::string weight_key) {
+    auto iter = weights_mem_map.find(weight_key);
+    if (iter != weights_mem_map.end())
+      return iter->second;
+    return nullptr;
+  }
+
+  void SetWeightsMemory(std::string weight_key,
+  std::shared_ptr<mkldnn::memory> filter_dst_mem) {
+      weights_mem_map[weight_key] = filter_dst_mem;
+  }
+
+ private:
+  // mkldnn weights(filer data) memory blocks from first iteration
+  // saved by weights name
+  std::map<std::string, std::shared_ptr<mkldnn::memory>> weights_mem_map;
```

Comment:
You don't need shared_ptr's. You can use unique_ptr and return raw ptrs.

Reply:
@pranavsharma @jywu-msft 

Tried the following and it works. But I had to

Remove forward_reference and include mkldnn.hpp. This will cause pybuild errors
```
namespace mkldnn {
class memory;
};
```

Convert raw-pointer to unique_ptr to Save memory in provider (conv.cc)
`std::unique_ptr<mkldnn::memory>{filter_dst_mem}`

const and reference doesn't work in SetWeightsMemory (mkldnn_provider.h)
    `weights_mem_map1_.emplace(weight_key, std::move(filter_dst_mem));`

Please suggest.

**mkldnn_provider.h**
```
  mkldnn::memory* GetWeightsMemory1(const std::string& weight_key) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto iter = weights_mem_map1_.find(weight_key);
    if (iter != weights_mem_map1_.end())
      return iter->second.get();
    return nullptr;
  }

  void SetWeightsMemory1(std::string weight_key,
                        std::unique_ptr<mkldnn::memory> filter_dst_mem) {
    std::lock_guard<std::mutex> lock(mutex_);
    weights_mem_map1_.emplace(weight_key, std::move(filter_dst_mem));
  }
  
 private:
  std::map<std::string, std::unique_ptr<mkldnn::memory>> weights_mem_map1_;

```
**conv.cc**
```
    mkldnn::memory* filter_dst_mem = nullptr;
    filter_dst_mem = provider_->GetWeightsMemory1(weight_name);

   if (filter_dst_mem == nullptr) {
      if (filter_format != conv_primitive->GetFilterMemoryFormat()) {
        auto pd = mkldnn::memory::primitive_desc(mkldnn::memory::desc(
                                                     filter_dims_mkl, MklDnnType<T>(), filter_format),
                                                 cpu_engine);
        mkldnn::memory src = mkldnn::memory(pd, (void*)filter_data);
        filter_reorder_buffer = IAllocator::MakeUniquePtr<void>(alloc, conv_primitive->GetFilterSize());
        filter_dst_mem = 
            new mkldnn::memory(conv_fwd_pd->weights_primitive_desc(), filter_reorder_buffer.get());
        MemoryReorderParams params(src, *filter_dst_mem);
        DoReorder<T>(params);
        provider_->SetWeightsMemory1(weight_name, std::unique_ptr<mkldnn::memory>{filter_dst_mem});
        filter_data = static_cast<T*>(filter_dst_mem->get_data_handle());
      }
    } else {
      filter_data = static_cast<T*>(filter_dst_mem->get_data_handle());
    }
```

Reply:
Why did you have to remove the forward reference? Doesn't keeping the forward ref work while using unique_ptr?

Reply:
I can not compile otherwise. :-(